22004 (#2190)
OpenLiberty/blogs: posts/2022-04-12-java18-22004.adoc

For more information on Java 18, please visit the Java 18 link:https://jdk.java.net/18/release-notes[release notes page], link:https://download.java.net/java/early_access/jdk18/docs/api/[API Javadoc page] or link:https://jdk.java.net/18/[download page].
For more information on Open Liberty, please visit our link:{url-prefix}/docs[documentation page].

[#certificate]
== Support client certificates with URL encoded strings

An `X.509` certificate is a digital certificate that is used to manage identity and security in internet communications and computer networking.

Intermediate servers, such as a web server, proxy server, or load balancer can send X.509 certificates to Open Liberty servers. Due to the deprecation of long lines folding in the `HTTP` header continuation, some open source intermediate servers might send a client certificate in the Privacy-Enhanced Mail (`PEM`) url-encoded format.

> Open Liberty 22.0.0.4 adds the ability to auto detect and process this new format.

Reviewer: ```suggestion
Open Liberty 22.0.0.4 adds the ability to automatically detect and process `X.509` certificates that are sent in PEM format.
```

An example of a truncated url-encoded certificate sent to application server:app-name:

[source, role="no_copy"]
----
[-----BEGIN%20CERTIFICATE-----%0AMIIDODCCAiCgAwIBAgIIHbO76YfCH1cwDQYJKoZIhvcNAQELBQAw
OjELMAkGA1UE%0ABhMCVVMxDDAKBgNVBAoTA0lCTTEMMAoGA1UECxMDVFdTMQ8wDQYDVQQDEwZDbGll%0AbnQ
wHhcNMTUxMTAzMTYyODM1WhcNMzUxMDMwMTYyODM1WjA6MQswCQYDVQQGEwJV%0AUzEMMAoGA1UEChMDSUJNM
QwwCgYDVQQLEwNUV1MxDz%2F%2Bpo%2FF%2Bzi%2F0sfAUwRdfgMm%2FAKAGzwMQufOYeKCgMULtq14QAJQnL
mq4M%2FM00%0AC5QyeYtdaTou%2BMsLmoa1tkq2VSDVxAcktJyRSRsox36G7EHDLV4U2gtR6xczNEjw%0A2%2
Bsj772FjdAMXRSR%0A-----END%20CERTIFICATE-----%0A]
----

Note: 

* (1) The header (i.e BEGIN ) and footer (i.e END) information are appended before sending to application server.

> * (2) The encoded character (i.e `%`) presents in the certificate.

Reviewer: what does this mean? presents what in the certificate? Does this refer specifically to `%` ? The statement suggests that the percent symbol is just one example of an encoded character but there's not enough information to understand the context or why this is important to note

Reviewer 2: @pmd1nh would you be able to clarify this for us so that we can make the appropriate change? Thanks in advance :)

Reviewer 3: @ryan-storey If the certificate string contains any % character, it indicates that the string is encoded.

at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5049)
... 20 more
----
+
We've fixed this issue so that the `JSP` now returns to the client without a `NPE` even in file not found situations.

[#guides]
== New and updated guides since the previous release
As Open Liberty features and functionality continue to grow, we continue to add link:https://openliberty.io/guides/?search=new&key=tag[new guides to openliberty.io] on those topics to make their adoption as easy as possible.  Since our last release, we've added the following two guides:

* link:{url-prefix}/guides/openliberty-operator-intro.html[Deploying a microservice to Kubernetes by using Open Liberty Operator]
** Explore how to deploy a microservice to Kubernetes by using Open Liberty Operator.
* link:{url-prefix}/guides/openliberty-operator-openshift.html[Deploying a microservice to OpenShift 4 by using Open Liberty Operator]
** Explore how to deploy a microservice to Red Hat OpenShift 4 by using Open Liberty Operator.

> Existing guides also receive updates in order to address any reported bugs/issues, keep their content current, and expand what their topic covers.  Since our last release, we've updated the majority of the guides to align with MicroProfile 5 and Jakarta EE 9.  Furthermore, our cloud hosted versions of the guides have also received various improvements.  For further information on these updates, refer to the link:{url-prefix}/blog/2022/04/12/guides-updated-mp5-jakartaee9.html[Learn MicroProfile 5.0 and Jakarta EE 9.1 from our newly enhanced cloud-hosted guides] blog post.

Reviewer: ```suggestion
We also update existing guides to address any reported bugs/issues, keep their content current, and expand what their topics cover.  Since our last release, we've updated the majority of the guides to align with MicroProfile 5 and Jakarta EE 9.  Furthermore, our cloud hosted versions of the guides have also received various improvements.  For further information on these updates, refer to the link:{url-prefix}/blog/2022/04/12/guides-updated-mp5-jakartaee9.html[Learn MicroProfile 5.0 and Jakarta EE 9.1 from our newly enhanced cloud-hosted guides] blog post.
```

Intermediate servers, such as a web server, proxy server, or load balancer can send X.509 certificates to Open Liberty servers. Due to the deprecation of long lines folding in the `HTTP` header continuation, some open source intermediate servers might send a client certificate in the Privacy-Enhanced Mail (`PEM`) url-encoded format.

Open Liberty 22.0.0.4 adds the ability to automatically detect and process `X.509` certificates that are sent in PEM format.

An example of a truncated url-encoded certificate sent to application server:app-name:

[source, role="no_copy"]
----
[-----BEGIN%20CERTIFICATE-----%0AMIIDODCCAiCgAwIBAgIIHbO76YfCH1cwDQYJKoZIhvcNAQELBQAw
OjELMAkGA1UE%0ABhMCVVMxDDAKBgNVBAoTA0lCTTEMMAoGA1UECxMDVFdTMQ8wDQYDVQQDEwZDbGll%0AbnQ
wHhcNMTUxMTAzMTYyODM1WhcNMzUxMDMwMTYyODM1WjA6MQswCQYDVQQGEwJV%0AUzEMMAoGA1UEChMDSUJNM
QwwCgYDVQQLEwNUV1MxDz%2F%2Bpo%2FF%2Bzi%2F0sfAUwRdfgMm%2FAKAGzwMQufOYeKCgMULtq14QAJQnL
mq4M%2FM00%0AC5QyeYtdaTou%2BMsLmoa1tkq2VSDVxAcktJyRSRsox36G7EHDLV4U2gtR6xczNEjw%0A2%2
Bsj772FjdAMXRSR%0A-----END%20CERTIFICATE-----%0A]
----

Note: 

* (1) The header (i.e BEGIN ) and footer (i.e END) information are appended before sending to application server.

> * (2) The encoded character (i.e `%`) presents in the certificate.

Reviewer: ```suggestion
* (2) The presence of the percent character (`%`) in a certificate indicates that the string is encoded.
```

* <<certificate, Support client certificate with URL encoded strings>>
* <<CVEs, Security Vulnerability (CVE) Fixes>>
* <<bugs, Notable bug fixes>>
* <<guides, New and updated guides since the previous release>>

> * <<paketo, Introducing the Paketo Liberty Buildpack>>

Author: ```suggestion
* <<paketo, Introducing the Paketo Liberty Buildpack>>
* <<CVEs, Security Vulnerability (CVE) Fixes>>
* <<bugs, Notable bug fixes>>
```

Author: We don't need the link to the guides section, as it's called out in the sentence below the ToC (and guides are not part of the runtime)

* link:https://openliberty.io/guides/docker.html[Using Docker containers to develop microservices]
* link:https://openliberty.io/guides/bean-validation.html[Validating constraints with microservices]

> [#paketo]

Author: On second thought, can you move this as the third item in the blog (as shown in the suggestion for the Toc), before the CVE section.

== Introducing the Paketo Liberty Buildpack

> The Paketo Liberty Buildpack is a new method to deploy your applications. Paketo buildpacks transform your application source code into container images and easily keep them updated. The Paketo Liberty buildpack provides the Open Liberty runtime to a workflow that produces an Open Container Initiative (OCI) image that can run just about anywhere. There are many advantages to using this buildpack which are explained in more detail in the link:{url-prefix}/blog/2022/04/01/cloud-native-liberty-buildpack.html[standalone blog post].

Author: ```suggestion
The Paketo Liberty Buildpack is a new method to deploy your applications. Paketo buildpacks transform your application source code into container images and easily keep them updated. The Paketo Liberty buildpack provides the Open Liberty runtime to a workflow that produces an Open Container Initiative (OCI) image that can run just about anywhere. There are many advantages to using this buildpack which are explained in more detail in the link:{url-prefix}/blog/2022/04/01/cloud-native-liberty-buildpack.html[Introducing the Paketo Liberty Buildpack] blog post.
```
